autoblocks/_impl/testing/run.py

```diff
@@ -0,0 +1,301 @@
+import asyncio
+import contextvars
+import dataclasses
+import inspect
+import threading
+import traceback
+from typing import Any
+from typing import Callable
+from typing import Coroutine
+from typing import List
+from typing import Optional
+
+import httpx
+
+from autoblocks._impl.testing.models import BaseEvaluator
+from autoblocks._impl.testing.models import BaseTestCase
+from autoblocks._impl.testing.models import Evaluation
+from autoblocks._impl.util import AutoblocksEnvVar
+
+# Globals
+client: httpx.AsyncClient = None
+loop: asyncio.AbstractEventLoop = None
+started: bool = False
+
+
+# Context
+current_test_id_var = contextvars.ContextVar("current_test_id")
+current_test_case_var = contextvars.ContextVar("current_test_case")
+
+
+def run_event_loop(_loop: asyncio.AbstractEventLoop) -> None:
+    asyncio.set_event_loop(_loop)
+    _loop.run_forever()
+
+
+def init_global_vars() -> None:
+    global client, loop, started
+
+    if started:
+        return
+
+    cli_server_address = AutoblocksEnvVar.CLI_SERVER_ADDRESS.get()
+    if not cli_server_address:
+        raise RuntimeError(
+            "Autoblocks tests must be run within the context of the testing CLI.\n"
+            "Make sure you are running your test command with:\n"
+            "$ npx autoblocks testing exec -- <your test command>"
+        )
+    client = httpx.AsyncClient(base_url=cli_server_address)
+
+    loop = asyncio.new_event_loop()
+
+    background_thread = threading.Thread(
+        target=run_event_loop,
+        args=(loop,),
+        daemon=True,
+    )
+    background_thread.start()
+
+    started = True
+
+
+async def send_error(
+    test_id: str,
+    test_case_hash: Optional[str],
+    evaluator_id: Optional[str],
+    error: Exception,
+) -> None:
+    await client.post(
+        "/errors",
+        json=dict(
+            testExternalId=test_id,
+            testCaseHash=test_case_hash,
+            evaluatorExternalId=evaluator_id,
+            error=dict(
+                name=type(error).__name__,
+                message=str(error),
+                stacktrace=traceback.format_exc(),
+            ),
+        ),
+    )
+
+
+async def gather_with_max_concurrency(
+    max_concurrency: int,
+    *coroutines: Coroutine,
```

Comment:
I didn't know you can use this `*` syntax here. 

Couldn't this also be `coroutines: list[Coroutine]`?

Reply:
yes, just wanted to match the original signature. but we can do that instead